File: packages/capacitor-notificare-authentication/ios/Plugin/NotificareAuthenticationPlugin.swift
import Foundation
import Capacitor
import NotificareKit
import NotificareAuthenticationKit

@objc(NotificareAuthenticationPlugin)
public class NotificareAuthenticationPlugin: CAPPlugin {
    
    public override func load() {
        EventBroker.instance.setup { self.notifyListeners($0, data: $1) }
        
        NotificationCenter.default.addObserver(self, selector: #selector(self.handleUniversalLink(notification:)), name: .capacitorOpenUniversalLink, object: nil)
    }
    
    deinit {
        NotificationCenter.default.removeObserver(self, name: .capacitorOpenUniversalLink, object: nil)
    }
    
    @objc func handleUniversalLink(notification: NSNotification) {
        guard let object = notification.object as? [String: Any?], let url = object["url"] as? URL else {
            NotificareLogger.warning("Unprocessable universal_link event.")
            return
        }
        
        if let token = Notificare.shared.authentication().parsePasswordResetToken(url) {
            EventBroker.instance.dispatchEvent("password_reset_token_received", data: [
                "token": token
            ])
            return
        }
        
        if let token = Notificare.shared.authentication().parseValidateUserToken(url) {
            EventBroker.instance.dispatchEvent("validate_user_token_received", data: [
                "token": token
            ])
            return
        }
    }
    
    // MARK: - Notificare Authentication
    
    @objc func isLoggedIn(_ call: CAPPluginCall) {
        call.resolve([
            "result": Notificare.shared.authentication().isLoggedIn
        ])
    }
    
    @objc func login(_ call: CAPPluginCall) {
        guard let email = call.getString("email") else {
            call.reject("Missing 'email' parameter.")
            return
        }
        
        guard let password = call.getString("password") else {
            call.reject("Missing 'password' parameter.")
            return
        }
        
        Notificare.shared.authentication().login(email: email, password: password) { result in
            switch result {
            case .success:
                call.resolve()
            case let .failure(error):
                call.reject(error.localizedDescription)
            }
        }
    }
    
    @objc func logout(_ call: CAPPluginCall) {
        Notificare.shared.authentication().logout { result in
            switch result {
            case .success:
                call.resolve()
            case let .failure(error):
                call.reject(error.localizedDescription)
            }
        }
    }
    
    @objc func fetchUserDetails(_ call: CAPPluginCall) {
        Notificare.shared.authentication().fetchUserDetails { result in
            switch result {
            case let .success(user):
                do {
                    call.resolve([
                        "result": try user.toJson()
                    ])
                } catch {
                    call.reject(error.localizedDescription)
                }
            case let .failure(error):
                call.reject(error.localizedDescription)
            }
        }
    }
    
    @objc func changePassword(_ call: CAPPluginCall) {
        guard let password = call.getString("password") else {
            call.reject("Missing 'password' parameter.")
            return
        }
        
        Notificare.shared.authentication().changePassword(password) { result in
            switch result {
            case .success:
                call.resolve()
            case let .failure(error):
                call.reject(error.localizedDescription)
            }
        }
    }
    
    @objc func generatePushEmailAddress(_ call: CAPPluginCall) {
        Notificare.shared.authentication().generatePushEmailAddress { result in
            switch result {
            case let .success(user):
                do {
                    call.resolve([
                        "result": try user.toJson()
                    ])
                } catch {
                    call.reject(error.localizedDescription)
                }
            case let .failure(error):
                call.reject(error.localizedDescription)
            }
        }
    }
    
    @objc func createAccount(_ call: CAPPluginCall) {
        guard let email = call.getString("email") else {
            call.reject("Missing 'email' parameter.")
            return
        }
        
        guard let password = call.getString("password") else {
            call.reject("Missing 'password' parameter.")
            return
        }
        
        let name = call.getString("name")
        
        Notificare.shared.authentication().createAccount(email: email, password: password, name: name) { result in
            switch result {
            case .success:
                call.resolve()
            case let .failure(error):
                call.reject(error.localizedDescription)
            }
        }
    }
    
    @objc func validateUser(_ call: CAPPluginCall) {
        guard let token = call.getString("token") else {
            call.reject("Missing 'token' parameter.")
            return
        }
        
        Notificare.shared.authentication().validateUser(token: token) { result in
            switch result {
            case .success:
                call.resolve()
            case let .failure(error):
                call.reject(error.localizedDescription)
            }
        }
    }
    
    @objc func sendPasswordReset(_ call: CAPPluginCall) {
        guard let email = call.getString("email") else {
            call.reject("Missing 'email' parameter.")
            return
        }
        
        Notificare.shared.authentication().sendPasswordReset(email: email) { result in
            switch result {
            case .success:
                call.resolve()
            case let .failure(error):
                call.reject(error.localizedDescription)
            }
        }
    }
    
    @objc func resetPassword(_ call: CAPPluginCall) {
        guard let password = call.getString("password") else {
            call.reject("Missing 'password' parameter.")
            return
        }
        
        guard let token = call.getString("token") else {
            call.reject("Missing 'token' parameter.")
            return
        }
        
        Notificare.shared.authentication().resetPassword(password, token: token) { result in
            switch result {
            case .success:
                call.resolve()
            case let .failure(error):
                call.reject(error.localizedDescription)
            }
        }
    }
    
    @objc func fetchUserPreferences(_ call: CAPPluginCall) {
        Notificare.shared.authentication().fetchUserPreferences { result in
            switch result {
            case let .success(preferences):
                do {
                    call.resolve([
                        "result": try preferences.map { try $0.toJson() }
                    ])
                } catch {
                    call.reject(error.localizedDescription)
                }
            case let .failure(error):
                call.reject(error.localizedDescription)
            }
        }
    }
    
    @objc func fetchUserSegments(_ call: CAPPluginCall) {
        Notificare.shared.authentication().fetchUserSegments { result in
            switch result {
            case let .success(segments):
                do {
                    call.resolve([
                        "result": try segments.map { try $0.toJson() }
                    ])
                } catch {
                    call.reject(error.localizedDescription)
                }
            case let .failure(error):
                call.reject(error.localizedDescription)
            }
        }
    }
    
    @objc func addUserSegment(_ call: CAPPluginCall) {
        guard let json = call.getObject("segment") else {
            call.reject("Missing 'segment' parameter.")
            return
        }
        
        let segment: NotificareUserSegment
        
        do {
            segment = try NotificareUserSegment.fromJson(json: json)
        } catch {
            call.reject(error.localizedDescription)
            return
        }
        
        Notificare.shared.authentication().addUserSegment(segment) { result in
            switch result {
            case .success:
                call.resolve()
            case let .failure(error):
                call.reject(error.localizedDescription)
            }
        }
    }
    
    @objc func removeUserSegment(_ call: CAPPluginCall) {
        guard let json = call.getObject("segment") else {
            call.reject("Missing 'segment' parameter.")
            return
        }
        
        let segment: NotificareUserSegment
        
        do {
            segment = try NotificareUserSegment.fromJson(json: json)
        } catch {
            call.reject(error.localizedDescription)
            return
        }
        
        Notificare.shared.authentication().removeUserSegment(segment) { result in
            switch result {
            case .success:
                call.resolve()
            case let .failure(error):
                call.reject(error.localizedDescription)
            }
        }
    }
    
    @objc func addUserSegmentToPreference(_ call: CAPPluginCall) {
        guard let preferenceJson = call.getObject("preference") else {
            call.reject("Missing 'preference' parameter.")
            return
        }
        
        guard let segmentJson = call.getObject("segment") else {
            call.reject("Missing 'segment' paramenter.")
            return
        }
        
        let preference: NotificareUserPreference
        
        do {
            preference = try NotificareUserPreference.fromJson(json: preferenceJson)
        } catch {
            call.reject(error.localizedDescription)
            return
        }
        
        do {
            let segment = try NotificareUserSegment.fromJson(json: segmentJson)
            Notificare.shared.authentication().addUserSegmentToPreference(segment, to: preference) { result in
                switch result {
                case .success:
                    call.resolve()
                case let .failure(error):
                    call.reject(error.localizedDescription)
                }
            }
            
            return
        } catch {
            NotificareLogger.debug("Failed to parse segment data into NotificareUserSegment.", error: error)
        }
        
        do {
            let option = try NotificareUserPreference.Option.fromJson(json: segmentJson)
            Notificare.shared.authentication().addUserSegmentToPreference(option: option, to: preference) { result in
                switch result {
                case .success:
                    call.resolve()
                case let .failure(error):
                    call.reject(error.localizedDescription)
                }
            }
            
            return
        } catch {
            NotificareLogger.debug("Failed to parse segment data into NotificareUserPreference.Option.", error: error)
        }
        
        call.reject("To execute this method, you must provide either a NotificareUserSegment or a NotificarePreferenceOption.")
    }
    
    @objc func removeUserSegmentFromPreference(_ call: CAPPluginCall) {
        guard let preferenceJson = call.getObject("preference") else {
            call.reject("Missing 'preference' parameter.")
            return
        }
        
        guard let segmentJson = call.getObject("segment") else {
            call.reject("Missing 'segment' paramenter.")
            return
        }
        
        let preference: NotificareUserPreference
        
        do {
            preference = try NotificareUserPreference.fromJson(json: preferenceJson)
        } catch {
            call.reject(error.localizedDescription)
            return
        }
        
        do {
            let segment = try NotificareUserSegment.fromJson(json: segmentJson)
            Notificare.shared.authentication().removeUserSegmentFromPreference(segment, from: preference) { result in
                switch result {
                case .success:
                    call.resolve()
                case let .failure(error):
                    call.reject(error.localizedDescription)
                }
            }
            
            return
        } catch {
            NotificareLogger.debug("Failed to parse segment data into NotificareUserSegment.", error: error)
        }
        
        do {
            let option = try NotificareUserPreference.Option.fromJson(json: segmentJson)
            Notificare.shared.authentication().removeUserSegmentFromPreference(option: option, from: preference) { result in
                switch result {
                case .success:
                    call.resolve()
                case let .failure(error):
                    call.reject(error.localizedDescription)
                }
            }
            
            return
        } catch {
            NotificareLogger.debug("Failed to parse segment data into NotificareUserPreference.Option.", error: error)
        }
        
        call.reject("To execute this method, you must provide either a NotificareUserSegment or a NotificarePreferenceOption.")
    }
}

File: packages/capacitor-notificare-authentication/android/src/main/java/re/notifica/authentication/capacitor/NotificareAuthenticationPlugin.kt
package re.notifica.authentication.capacitor

import android.content.Intent
import com.getcapacitor.*
import com.getcapacitor.annotation.CapacitorPlugin
import re.notifica.Notificare
import re.notifica.NotificareCallback
import re.notifica.authentication.ktx.authentication
import re.notifica.authentication.models.NotificareUser
import re.notifica.authentication.models.NotificareUserPreference
import re.notifica.authentication.models.NotificareUserSegment
import re.notifica.internal.NotificareLogger

@CapacitorPlugin(name = "NotificareAuthenticationPlugin")
public class NotificareAuthenticationPlugin : Plugin() {

    override fun load() {
        EventBroker.setup(this::notifyListeners)

        val intent = activity?.intent
        if (intent != null) handleOnNewIntent(intent)
    }

    override fun handleOnNewIntent(intent: Intent) {
        val passwordResetToken = Notificare.authentication().parsePasswordResetToken(intent)
        if (passwordResetToken != null) {
            EventBroker.dispatchEvent("password_reset_token_received", JSObject().apply {
                put("token", passwordResetToken)
            })
            return
        }

        val validateUserToken = Notificare.authentication().parseValidateUserToken(intent)
        if (validateUserToken != null) {
            EventBroker.dispatchEvent("validate_user_token_received", JSObject().apply {
                put("token", validateUserToken)
            })
            return
        }
    }

    @PluginMethod
    public fun isLoggedIn(call: PluginCall) {
        call.resolve(
            JSObject().apply {
                put("result", Notificare.authentication().isLoggedIn)
            }
        )
    }

    @PluginMethod
    public fun login(call: PluginCall) {
        val email = call.getString("email") ?: run {
            call.reject("Missing 'email' parameter.")
            return
        }

        val password = call.getString("password") ?: run {
            call.reject("Missing 'password' parameter.")
            return
        }

        Notificare.authentication().login(email, password, object : NotificareCallback<Unit> {
            override fun onSuccess(result: Unit) {
                call.resolve()
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun logout(call: PluginCall) {
        Notificare.authentication().logout(object : NotificareCallback<Unit> {
            override fun onSuccess(result: Unit) {
                call.resolve()
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun fetchUserDetails(call: PluginCall) {
        Notificare.authentication().fetchUserDetails(object : NotificareCallback<NotificareUser> {
            override fun onSuccess(result: NotificareUser) {
                try {
                    call.resolve(
                        JSObject().apply {
                            put("result", result.toJson())
                        }
                    )
                } catch (e: Exception) {
                    call.reject(e.localizedMessage)
                }
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun changePassword(call: PluginCall) {
        val password = call.getString("password") ?: run {
            call.reject("Missing 'password' parameter.")
            return
        }

        Notificare.authentication().changePassword(password, object : NotificareCallback<Unit> {
            override fun onSuccess(result: Unit) {
                call.resolve()
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun generatePushEmailAddress(call: PluginCall) {
        Notificare.authentication().generatePushEmailAddress(object : NotificareCallback<NotificareUser> {
            override fun onSuccess(result: NotificareUser) {
                try {
                    call.resolve(
                        JSObject().apply {
                            put("result", result.toJson())
                        }
                    )
                } catch (e: Exception) {
                    call.reject(e.localizedMessage)
                }
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun createAccount(call: PluginCall) {
        val email = call.getString("email") ?: run {
            call.reject("Missing 'email' parameter.")
            return
        }

        val password = call.getString("password") ?: run {
            call.reject("Missing 'password' parameter.")
            return
        }

        val name = call.getString("name")

        Notificare.authentication().createAccount(email, password, name, object : NotificareCallback<Unit> {
            override fun onSuccess(result: Unit) {
                call.resolve()
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun validateUser(call: PluginCall) {
        val token = call.getString("token") ?: run {
            call.reject("Missing 'token' parameter.")
            return
        }

        Notificare.authentication().validateUser(token, object : NotificareCallback<Unit> {
            override fun onSuccess(result: Unit) {
                call.resolve()
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun sendPasswordReset(call: PluginCall) {
        val email = call.getString("email") ?: run {
            call.reject("Missing 'email' parameter.")
            return
        }

        Notificare.authentication().sendPasswordReset(email, object : NotificareCallback<Unit> {
            override fun onSuccess(result: Unit) {
                call.resolve()
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun resetPassword(call: PluginCall) {
        val password = call.getString("password") ?: run {
            call.reject("Missing 'password' parameter.")
            return
        }

        val token = call.getString("token") ?: run {
            call.reject("Missing 'token' parameter.")
            return
        }

        Notificare.authentication().resetPassword(password, token, object : NotificareCallback<Unit> {
            override fun onSuccess(result: Unit) {
                call.resolve()
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun fetchUserPreferences(call: PluginCall) {
        Notificare.authentication().fetchUserPreferences(object : NotificareCallback<List<NotificareUserPreference>> {
            override fun onSuccess(result: List<NotificareUserPreference>) {
                try {
                    call.resolve(
                        JSObject().apply {
                            put("result", JSArray(result.map { it.toJson() }))
                        }
                    )
                } catch (e: Exception) {
                    call.reject(e.localizedMessage)
                }
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun fetchUserSegments(call: PluginCall) {
        Notificare.authentication().fetchUserSegments(object : NotificareCallback<List<NotificareUserSegment>> {
            override fun onSuccess(result: List<NotificareUserSegment>) {
                try {
                    call.resolve(
                        JSObject().apply {
                            put("result", JSArray(result.map { it.toJson() }))
                        }
                    )
                } catch (e: Exception) {
                    call.reject(e.localizedMessage)
                }
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun addUserSegment(call: PluginCall) {
        val json = call.getObject("segment") ?: run {
            call.reject("Missing 'segment' parameter.")
            return
        }

        val segment: NotificareUserSegment = try {
            NotificareUserSegment.fromJson(json)
        } catch (e: Exception) {
            call.reject(e.localizedMessage)
            return
        }

        Notificare.authentication().addUserSegment(segment, object : NotificareCallback<Unit> {
            override fun onSuccess(result: Unit) {
                call.resolve()
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun removeUserSegment(call: PluginCall) {
        val json = call.getObject("segment") ?: run {
            call.reject("Missing 'segment' parameter.")
            return
        }

        val segment: NotificareUserSegment = try {
            NotificareUserSegment.fromJson(json)
        } catch (e: Exception) {
            call.reject(e.localizedMessage)
            return
        }

        Notificare.authentication().removeUserSegment(segment, object : NotificareCallback<Unit> {
            override fun onSuccess(result: Unit) {
                call.resolve()
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })

    }

    @PluginMethod
    public fun addUserSegmentToPreference(call: PluginCall) {
        val preferenceJson = call.getObject("preference") ?: run {
            call.reject("Missing 'preference' parameter.")
            return
        }

        val segmentJson = call.getObject("segment") ?: run {
            call.reject("Missing 'segment' parameter.")
            return
        }

        val preference: NotificareUserPreference = try {
            NotificareUserPreference.fromJson(preferenceJson)
        } catch (e: Exception) {
            call.reject(e.localizedMessage)
            return
        }

        try {
            val segment = NotificareUserSegment.fromJson(segmentJson)
            Notificare.authentication()
                .addUserSegmentToPreference(segment, preference, object : NotificareCallback<Unit> {
                    override fun onSuccess(result: Unit) {
                        call.resolve()
                    }

                    override fun onFailure(e: Exception) {
                        call.reject(e.localizedMessage)
                    }
                })

            return
        } catch (e: Exception) {
            NotificareLogger.debug("Failed to parse segment data into a NotificareUserSegment.", e)
        }

        try {
            val option = NotificareUserPreference.Option.fromJson(segmentJson)
            Notificare.authentication()
                .addUserSegmentToPreference(option, preference, object : NotificareCallback<Unit> {
                    override fun onSuccess(result: Unit) {
                        call.resolve()
                    }

                    override fun onFailure(e: Exception) {
                        call.reject(e.localizedMessage)
                    }
                })

            return
        } catch (e: Exception) {
            NotificareLogger.debug("Failed to parse segment data into a NotificareUserPreference.Option.", e)
        }

        call.reject("To execute this method, you must provide either a NotificareUserSegment or a NotificarePreferenceOption.")
    }

    @PluginMethod
    public fun removeUserSegmentFromPreference(call: PluginCall) {
        val preferenceJson = call.getObject("preference") ?: run {
            call.reject("Missing 'preference' parameter.")
            return
        }

        val segmentJson = call.getObject("segment") ?: run {
            call.reject("Missing 'segment' parameter.")
            return
        }

        val preference: NotificareUserPreference = try {
            NotificareUserPreference.fromJson(preferenceJson)
        } catch (e: Exception) {
            call.reject(e.localizedMessage)
            return
        }

        try {
            val segment = NotificareUserSegment.fromJson(segmentJson)
            Notificare.authentication()
                .removeUserSegmentFromPreference(segment, preference, object : NotificareCallback<Unit> {
                    override fun onSuccess(result: Unit) {
                        call.resolve()
                    }

                    override fun onFailure(e: Exception) {
                        call.reject(e.localizedMessage)
                    }
                })

            return
        } catch (e: Exception) {
            NotificareLogger.debug("Failed to parse segment data into a NotificareUserSegment.", e)
        }

        try {
            val option = NotificareUserPreference.Option.fromJson(segmentJson)
            Notificare.authentication()
                .removeUserSegmentFromPreference(option, preference, object : NotificareCallback<Unit> {
                    override fun onSuccess(result: Unit) {
                        call.resolve()
                    }

                    override fun onFailure(e: Exception) {
                        call.reject(e.localizedMessage)
                    }
                })

            return
        } catch (e: Exception) {
            NotificareLogger.debug("Failed to parse segment data into a NotificareUserPreference.Option.", e)
        }

        call.reject("To execute this method, you must provide either a NotificareUserSegment or a NotificarePreferenceOption.")
    }
}
